components/org.wso2.carbon.identity.authenticator.saml2.sso/src/main/java/org/wso2/carbon/identity/authenticator/saml2/sso/SAML2SSOAuthenticator.java

```diff
@@ -790,23 +790,30 @@ private String generatePassword(String username) {
                 for (Attribute attribute : attributesList) {
                     String attributeName = attribute.getName();
                     if (attributeName != null && roleClaim.equals(attributeName)) {
-                        // Assumes role claim appear only once
-                        Element value = attribute.getAttributeValues().get(0).getDOM();
-                        String attributeValue = value.getTextContent();
-
-                        if (log.isDebugEnabled()) {
-                            log.debug("AttributeName : " + attributeName + ", AttributeValue : " + attributeValue);
-                        }
-
-                        roles = attributeValue.split(getAttributeSeperator());
-                        if (log.isDebugEnabled()) {
-                            log.debug("Role list : " + Arrays.toString(roles));
+                        List<XMLObject> attributeValues = attribute.getAttributeValues();
+                        if (attributeValues != null && attributeValues.size() == 1) {
+                            String attributeValueString = ((XSStringImpl) attributeValues.get(0)).getValue();
+                            String multiAttributeSeparator = getAttributeSeperator();
+                            String[] attributeValuesArray = attributeValueString.split(multiAttributeSeparator);
+                            roles.addAll(Arrays.asList(attributeValuesArray));
+                        } else if (attributeValues != null && attributeValues.size() > 1) {
+                            for (XMLObject attributeValue : attributeValues) {
+                                String attributeValueString = ((XSStringImpl) attributeValue).getValue();
+                                if (log.isDebugEnabled()) {
+                                    log.debug("AttributeName : " + attributeName + ", AttributeValue : " +
```

Comment:
Can improve log, "Adding attributes for Assertion: "+assertion+" AttributeName : " + attributeName + ...